### ISSUE TITLE
Expandable grid display issues when rows are expanded

### ISSUE THREAD

**Issue by yurigenin** [angular-ui/ui-grid#4508]
When you expand several top rows in the grid and scroll all the way down the grid is not painted properly. Just try Expandable Grid tutorial and you will see the problem. Seems like a scrolling issue but currently this feature is not much usable. You can see even bigger problem when you expand all rows. Now scrolling to the bottom exposes an empty flickering grid.

**Comment by yurigenin** [comment]
Seems like it is related to the virtualizationThreshold feature. If you make it equal to the number of rows in the main grid the issue goes away but with big datasets it is impractical as grid really slows to a crawl in that case.

**Comment by yurigenin** [comment]
The issue is in the GridRenderContainer.prototype.adjustRows method (expanded rows heights are not used for calculations) as well as in the uiGridViewport directive's controller() function:

`var hiddenRowWidth = (rowContainer.currentTopRow) * rowContainer.grid.options.rowHeight;`

Again, expanded row heights are not taken into consideration at all.

**Comment by MangeshPimpalkar** [comment]
+1

**Comment by FlavorScape** [comment]
Indeed, I can reproduce with a tall expandable row template. 

**Heres a plunkr of the tutorial demo with a tall expandable row:**
http://plnkr.co/edit/kLMMl74UjRXiuS3h0s8x?p=preview

**Comment by vance** [comment]
Hey, this is blocking me on a project. I don't have time to learn the entire stack in the API, but I think I may have found where to work on this. Seeing as this has been open for a few months, I may take a crack at it. I suspect it may be in the function `adjustScroll` due to this comment:

```
       * If we're scrolling up we scroll to the first row of the old data set -
       * so we're assuming that you would have gotten to the top of the grid (from the 20% need more data trigger) by
       * the time the data comes back.  If we're scrolling down we scoll to the last row of the old data set - so we're
       * assuming that you would have gotten to the bottom of the grid (from the 80% need more data trigger) by the time
       * the data comes back.
       *
       * Neither of these are good assumptions, but making this a smoother experience really requires
       * that trigger to not be a percentage, and to be much closer to the end of the data (say, 5 rows off the end).  Even then
       * it'd be better still to actually run into the end.  But if the data takes a while to come back, they may have scrolled
       * somewhere else in the mean-time, in which case they'll get a jump back to the new data.  Anyway, this will do for
       * now, until someone wants to do better.
       * @param {Grid} grid the grid we're working on
```

and the part of the code in question that I will examine with scrutiny is: 
 ```
newTop = grid.infiniteScroll.prevScrollTop || (grid.infiniteScroll.previousVisibleRows*rowHeight - viewportHeight);
```

As I would guess that the calculated height is inaccurate here.

**Comment by PaulL1** [comment]
The defect we're on here is with expandable, the code fragment you refer to is in infinite scroll.  So I doubt the two are related - the features are quite independent of each other.

**Comment by vance** [comment]
Yeah, just realized that. I'm spelunking around Grid class right now to get my bearings. Just seeing what normal height calculations are so I can compare it to the computed height of expanded rows.

**Comment by vance** [comment]
Ok, so back to this today. Will post more soon. In my initial research, I see one value that jumps to a very invalid value when scrolling above and below an expanded row. this alters the scrollPercent

GridRenderContainer.js
``` 
GridRenderContainer.prototype.adjustRows = function adjustRows(scrollTop, scrollPercentage, postDataLoaded) {
         ...
        self.getVerticalScrollLength()
        ...
```
returns 14992 constantly, but while scrolling will occasionally return **-215** (negative value)

**Comment by vance** [comment]
Here is a graph of the variables i've been watching. The undulation of the values scrolling up and down past the expanded row. There's some fishy stuff here. Note the jumping of the row index value by a huge margin. this seems to coincide with the jump in the verticalScrollLength value, sometimes being normal, sometimes bing off until the anomaly happens again.  
![ui-grid-graph-cropped](https://cloud.githubusercontent.com/assets/145471/15001823/aa145c06-114f-11e6-9ca8-ff052d43e309.jpg)

**Comment by vance** [comment]
I have graphed some variables and added the ability to tag events. We can clearly see the the parts where the yellow line (canvasHeight) jumps to a weird value when scrolling past the expanded row. This may be removing and adding it to the rowCache... but as you can see, there's two values it jumps to, so i'm thinking maybe it's both hitting and not hitting the `self.canvasHeightShouldUpdate`

<img width="1381" alt="screen shot 2016-05-04 at 11 08 57 am" src="https://cloud.githubusercontent.com/assets/145471/15024175/b11b2d24-11e8-11e6-8b64-4810a9bf6cea.png">

**Comment by vance** [comment]
Getting closer. I think there's an outstanding task to actually take height into account when determining the minRowsToRender... When I log if the rows have a height over 30  (row default height) it never logs true. So, the list is wrong, or the values are incorrect and do not include the expanded row's height. This is only one place... this may be lacking in other places, too....

```
GridRenderContainer.prototype.minRowsToRender = function minRowsToRender() {
    var self = this;
    var minRows = 0;
    var rowAddedHeight = 0;
    var viewPortHeight = self.getViewportHeight();
    var hasTall = false;
    for (var i = self.visibleRowCache.length - 1; rowAddedHeight < viewPortHeight && i >= 0; i--) {
      rowAddedHeight += self.visibleRowCache[i].height;    
      if(self.visibleRowCache[i].height > 30){
        hasTall = true;
      }
      minRows++;
    }
    console.log('has a tall row? ', hasTall );
    
    return minRows;
  };
```

**Comment by vance** [comment]
I'm going nuts here... anyone know where the visibleRowCache is actually set? All I can find are references getting the value, but not setting it. Maybe i'm blind.

**Comment by vance** [comment]
Of course, I find it right after posting: `Grid.prototype.setVisibleRows`

**Comment by vance** [comment]
1) Does anyone think expanding a row should call `redrawInPlace` on the container? (it doesn't) Or should it just recalculate the canvas height *only*?

2) Any idea why `GridRenderContainer.prototype.getCanvasHeight` shows `self.visibleRowCache` to contain a row with height > 30 (the expanded row), while in the function `minRowsToRender` the cache does not contain the expanded row?

**Comment by vance** [comment]
**My most recent finding**
It is removing the expanded row TOO EARLY. The currentTopRow index is incorrect and does not include the expanded row when scrolling down past it. That explains the "jumping" to 5-10 rows below the index where the canvas should start rendering.

I think this logic inside adjustRows should be inclusive of row height (including expanded rows). Currently, it seems to only decide the rendering range based on index/count and not height. 
```

if (rowCache.length > self.grid.options.virtualizationThreshold) {
  if (!(typeof(scrollTop) === 'undefined' || scrollTop === null)) {
    // Have we hit the threshold going down?
    if ( !self.grid.suppressParentScrollDown && self.prevScrollTop < scrollTop && rowIndex < self.prevRowScrollIndex + self.grid.options.scrollThreshold && rowIndex < maxRowIndex) {
      return;
    }
    //Have we hit the threshold going up?
    if ( !self.grid.suppressParentScrollUp && self.prevScrollTop > scrollTop && rowIndex > self.prevRowScrollIndex - self.grid.options.scrollThreshold && rowIndex < maxRowIndex) {
      return;
    }
  }
  var rangeStart = {};
  var rangeEnd = {};

  rangeStart = Math.max(0, rowIndex - self.grid.options.excessRows);
  rangeEnd = Math.min(rowCache.length, rowIndex + minRows + self.grid.options.excessRows);

  newRange = [rangeStart, rangeEnd];
}

```

**Comment by vance** [comment]
ah yes! Whilst scrolling past an expandable with height 1500PX, the `rowIndex` continues  to increment, even though we're still looking at the guts of the expandable row. it should be "locked" at the index of the expandable row until it has passed the top boundary of the canvas.

**Comment by vance** [comment]
Yay. I fixed half the problem! We need to compute the proper index based on where scrollTop lands compared to the expanded rows' positions.

doing this:
```
var theoreticalRowIndex = 0;
    var currentHeight = 0;
    var i = 0;
    self.visibleRowCache.forEach(function(r){
      if( scrollTop > currentHeight && scrollTop < currentHeight + r.height){
        theoreticalRowIndex = i;
      }
      currentHeight+= r.height;
      i++;
    });
``` 
instead of 
```
 maxRowIndex * scrollPercentage
```

gets rid of the "jumping" at the end of the expandable row. 

the new problem caused by this is blank rows after scrolling down for a bit, which is explainable: the end of the visible row index also needs to be corrected.

**Comment by vance** [comment]
hey my PR passes style, but I couldn't get the unit tests to run locally

**Comment by vance** [comment]
this is fixed with the PR by the way =)

**Comment by cutterbl** [comment]
I notice this PR was entered back in 2016. I implemented this locally (dc183a1), and it seems to resolve the scrolling issue with expandable rows. What's required to get the changes into code? I see a 'help wanted' tag here...

**Comment by mportuga** [comment]
@cutterbl  That PR id not valid, it is the equivalent of changing your grid options to have no virtual rows. See the final comment in the PR: https://github.com/angular-ui/ui-grid/pull/5395

**Comment by Agarwal-Pulkit** [comment]
I am also facing same issues while integrating expandable in our grid. Do we have any work around or solution that can eliminate scroll issue.
